### ISSUE TITLE
Allow State to be used in dependency lists for React.useEffect, React.useMemo, etc. and as prop for React.memo

### ISSUE THREAD

**Issue by magne4000** [avkonst/hookstate#140]
First of all, thank you for this awesome lib!

I found myself having [difficulties](https://github.com/avkonst/hookstate/issues/139) trying to find the right way to implement some state updates efficiently, and the following reflexion arose:

What about using Immer? Thanks to its support for [patches](https://immerjs.github.io/immer/docs/patches), would allow this lib to support a wider range of operations, on all sort of data types. `set` and `merge` methods could be replaced by `produce`, and plugins could leverage patches (which would be useful for [my plugin](https://github.com/getstation/hookstate-plugin-web-extension)).

I really like the whole intelligent proxy system that is currently in place, and I honestly think that putting Immer in there would make the whole lib even better!

I wonder if this is something possible? Or is there some limitations or willingness to keep everything inside hookstate itself?

**Comment by avkonst** [comment]
Hi. Thank you for the feedback. If you find the Hookstate helpful, please, spread few words about it in post or blogs to help the project grow.

Regarding your plugin.
It is unclear what your plugin intends to do exactly, but I think you might find this extension helpful to rely on: https://hookstate.js.org/docs/extensions-broadcasted (I open sourced it today)
I frequently combine this plugin with persistence to IndexDB in a leader tab or synchronization to a remote server.

Regarding Immer.
Immer was ruled out because it clones and copies data due to its immutable nature, which is slow and impacts the performance.
I see that sometimes it could be a case when you would like to revert changes (this is the main reason to use Immer patches, as far as I understand). Is it what you would like to achieve? If yes, you can do it like the following:

```
let state = useState({
    name: "Micheal",
    age: 32
});

// all the changes the user made in the wizard
let changes = []
// the inverse of all the changes made in the wizard
let inverseChanges = []

// apply changes
state.set(p => produce(
    p,
    draft => {
        draft.age = 33
    },
    // The third argument to produce is a callback to which the patches will be fed
    (patches, inversePatches) => {
        changes.push(...patches)
        inverseChanges.push(...inversePatches)
    }
))

// revert changes
state.set(p => applyPatches(p, inverseChanges))

```

Does it help?

**Comment by magne4000** [comment]
Thanks for your answer! The Broadcast plugin is quite useful indeed.

Quickly explanation of what my plugin does: it is a mix of the Broadcast plugin and the Persist plugin, but tailored for [Web Extension](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API). The 2 main points are:
- Use `browser.storage` instead of `localStorage` for persistence
- Also use `browser.storage` for [broadcasting](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/onChanged)

The whole point of using Immer in my case would be to ensure that any operation that I apply on a state only update what's necessary. If I'm limited by what I can do with `merge`, I should use `set` as you suggested.
But if I understand correctly how hookstate works, using `set` to insert one item in the middle of an array of 10000 items would not be efficient, as it will consider the whole array as a new one, and each of its elements also (please correct me if I'm wrong). And that's the kind of operations I need to do.

Last thing, using Immer own patch mecanism could be leveraged to replace internal `merged` + `path` computation, which is comprehensive and easy to broadcast and reapply operations onto another state.

**Comment by avkonst** [comment]
Ok. Let's assume that Produce function is available in Hookstate, and it captures exactly what parts of a state are patched, and so on. And it does it in such a way, that it takes control of triggering rerendering for only "patched" sub-trees of a state. Now consider the following component (it uses [StateFragment](https://hookstate.js.org/docs/using-without-statehook/) to opt-into scoped state ):

```tsx
function MyComponent() {
    let state = useState([1, 2, 3]);
    return <>
    {
        state.keys.map(i => <StateFragment key={i} state={state[i]}>{itemState => <p>{itemState.value}</p>}</StateFragment>)
    }
    </>
}
```

Now, you apply the following produce action:
```
let changes = Produce(state, draft => {
    delete draft[0]; // delete one element
    draft.push(5000); // insert new element
})
```

Now you need we captured changes done, precisely. And the "size" of the changes is proportional to changes done, and is independent from the length of the array state.
Next you need to come up with an answer: what components should be rerendered for a given set of captured changes. So, for the example above, what components should be rerendered?

**Comment by magne4000** [comment]
- `MyComponent` will rerender once because it has been updated
- All `MyItemComponent` will rerender because their values changed (even if for some their key did not)

In a different example, more suitable for what I'm trying to achieve:
```ts
function MyComponent() {
    let state = useState([
      { id: 1, value: someObject1 },
      { id: 2, value: someObject2 },
      { id: 3, value: someObject3 },
    ]);
    return <>
    {
        state.map((o, i) => <MyItemComponent key={o.id} state={state[i]} />)
    }
    </>
}

function MyItemComponent(props: {state: State<number>}) {
    let state = useState(props.state); // scoped state
    return <p>{state.value}</p>
}

...

let changes = Produce(state, draft => { 
    draft.splice(0, 1); // delete first element
    draft.push({ id: 4, someObject4 }); // insert new element
})
```

- `MyItemComponent[0]` should not rerender, `key` went from 1 to 2, and 2 was already rendered, and value did not change
- `MyItemComponent[1]` should not rerender, `key` went from 2 to 3, and 3 was already rendered, and value did not change
- `MyItemComponent[2]` should rerender, `key` was not existing before that

With that in mind, if I have to stream the changes that happened between process, I want those changes to reflects exactly the 2 actions that happened (delete first element, append one element to the list), otherwise I would stream the whole update array as changed, and thus it will set all 3 values to new references:

```ts
// --- Example with Immer patching
const changes = [
  { op: 'delete', path: [0] },
  { op: 'append', value: { id: 4, someObject4 } }
];

// Applying back those changes through Immer applyPatches will let my values with `id:2` and `id:3` untouched
// -> only 2 rerender, the parent and id:4

// --- Example with current Hookstate set (not actual format, but the whole array is streamed)
const changes = {
  set: [
    { id: 2, value: someObject2 },
    { id: 3, value: someObject3 },
    { id: 4, someObject4 }
  ]
}

// Apply back those changes will overwrite `id:2` and `id:3`, and they will rerender
// -> all components rerender
```

I hope this clears-up what I would like to achieve.

**Comment by avkonst** [comment]
Park streaming of changes for a moment. There are no technical barriers to come up with API over Immer which would allow to capture patches.

 The problem of rerendering is unfortunately harder than you see it now. Let's come back to your example. Keeping in mind these changes:
```
const changes = [
  { op: 'delete', path: [0] },
  { op: 'append', value: { id: 4, v: someObject4 } }
];
```
how would it not trigger rerendering for `MyComponent?` 

This operation `state.map(o => ...)` loops over keys (JS does it behind the map function) and the keys are changed. Somehow, the `MyComponent` should remove `MyItemComponent` children at first position and append new `MyItemComponent` at the last position.... I do not see how it can be done without asking the React to reread the state at MyComponent level.

**Comment by magne4000** [comment]
I forgot to mention that for the specific cases or elements that can changes places in a `map`, I also need to use `React.memo` to avoid unnecessary rerenders.
To better illustrate, I created a little codesanbox: https://codesandbox.io/s/relaxed-glitter-ytb2x
Both parts delete first element of an array of 20 elements, and edit the new first element to append a string, every seconds

You can see that React without dependencies is able to not rerender components that have just moved, but I'm unable to replicate this with hookstate

**Comment by avkonst** [comment]
Well, the behaviour is different because in hookstate case you memoise the state tracker for a value and react case you memoise the value of a state. To make things equivalent, it should be like the following:

```tsx
const CompCore = ({ o }) => {
  const val = useStateCore(o);

  return useMemo(() => 
    <li>
      Last render at: {new Date().toISOString()}. Value: {val.s.value}
    </li>,
    [val.s.value]
  );
};
```

I understand people would like to have stable state tracking object (ie. same `State` instance for the same `State.value` object), but it is hard (although I have got some ideas how it can be achieved). At very minimum `State` instances returned by 2 different `useState` calls should be different even for the same `State.value` object.

**Comment by avkonst** [comment]
I experimented with source code a bit and I think it should be possible to add support for memo as in your example, but it should be opt-in functionality as it will become non-backward compatible. It might even improve performance but it will require keeping references to State objects for longer (to be precise, it will require 2 re-renders to get rid of stale State objects related to deleted properties. Now they are cleaned up immediately on the first rerender)

**Comment by magne4000** [comment]
I just experienced that Immer patch generation is not that good since 7.0. In my example executing `.slice(0,1)` generates one replace operation per element of the array, instead of just one remove action, which complicates optimal syncing of state updates.
Nevertheless, if we could at least have hookstate not breaking `React.memo`, it would be a good step forward. I'm wondering if it is not possible to leverage the [`areEqual` second parameter of `React.memo`](https://reactjs.org/docs/react-api.html#reactmemo) at first, with pseudo code looking like this:
```ts
function MyComponent(props) {
  /* render using props */
}
function areEqual(prevProps, nextProps) {
  const props = nextProps.keys();
  return props.every(p => {
    if (nextProps[p] instanceof StateMethodsImpl) {
      // compare underlying real values
      return nextProps[p].realValue === prevProps[p].realValue;
    }
    // default behaviour
    return nextProps[p] === prevProps[p];
  });
}
export default React.memo(MyComponent, areEqual);
```

**Comment by avkonst** [comment]
I was thinking about the second parameter, but can not answer the question how to compare "real values": `nextProps[p].realValue === prevProps[p].realValue;`

I think it should be deep equality check for all of the deeply nested props, otherwise rerenders will be missed.
And once you do deep equality check you effectively mark the entire value and all its children and grand children being used at the time of `memo` call, which happens before scoped state is subscribed. It means the entire MyComponent will rerender (although not every MyItemComponent will be called because of the memo gatekeeping) when you update any prop of an item in the array....

I am happy to move this forward and do something about it. There are other similar tickets asking how to compare states in dependencies lists, like useMemo and useEffect. But unfortunately, I can not figure out how to compare values correctly? If we compare it in a such a way that a value is not marked "used enough" we will miss rerenders because of memo gatekeeping. If we mark it "well used" via deep equality comparison, we will kill the benefits of the scoped state. The later is simply a fallback to the the classic React useState, which is easily possible with Hookstate, like the following:

```tsx
function MyComponent() {
    let state = useState([
      { id: 1, v: someObject1 },
      { id: 2, v: someObject2 },
      { id: 3, v: someObject3 },
    ]);
    return <>
    {
        // notice Downgraded below, which is opt out from state usage tracking
        // and since the value is used the state is marked as used
        // and when any of its props are set, the root useState will trigger rerender
        state.map(o => <MyItemComponent key={Untracked(o.id).get()} state={o.v.attach(Downgraded).value} />)
    }
    </>
}

const MyItemComponent = memo((props: {stateValue:  { id: number, v: object}}) => {
    // use state value here, no access to setters and no scoped state
}, (prevProps, nextProps) => { /* compare here, maybe doing deep equality check */})
```

What do you think?

**Comment by avkonst** [comment]
I have done a bit of research and I think it should be possible to achieve something like the following:

```tsx
function MyComponent() {
    let state = useState([
      { id: 1, v: someObject1 },
      { id: 2, v: someObject2 },
      { id: 3, v: someObject3 },
    ]);
    let memorizedState = Hookstate.useMemo(state) // THIS will be either a new plugin or new extension to the Core.
    // memorizedState will be always the same object until the state or any of it's children are updated
    return <>
    {
        memorizedState.map(
            // NOTICE: o will be always the same object until the state[o.id] or any of its children are updated
            o => <MyItemComponent key={Untracked(o.id).get()} state={o.v} />)
    }
    </>
}

// NOTICE react memo here used with the state prop
const MyItemComponent = React.memo((props: {state: State<number>}) => {
    let state = useState(props.state); // scoped state
    return <p>{state.value}</p>
})
```

This would allow to use the state in dependencies lists for useEffect, useMemo, etc.. And React.memo would NOT require special `areEqual`.

The only problem is I am extremely busy with other projects at the moment, and I am not sure when I can get free time for this. This feature is quite large to complete within a quick hack session. I would need a couple of hours more to confirm that the idea is feasible and a couple of days to complete production ready implementation. I wonder if you could sponsor this development? then it would make it a lot easier to allocate proper time for this.

**Comment by avkonst** [comment]
Updated the proposed code sample after thinking a bit more.

**Comment by magne4000** [comment]
I like the memoizer approach, it seems to fit better into Hookstate philosophy.
With this in place + #146, this lib should be able to handle all kind of cumbersome cases. I'm quite busy too, but I'm willing to propose PRs or at least reviews if you need.
Also thanks for renaming the topic which 👌

**Comment by avkonst** [comment]
This feature has been finally implemented on a branch. Thanks for XRFoundation sponsoring the development. This will get merged to master and released in Hookstate 4 coming weeks.

**Comment by avkonst** [comment]
We have Release Candidate but this feature needs documentation update and few more things for useMemo useCallback hooks. Support for useEffect hook has been done. It depends on how soon I put enough effort to make it happen. I would say the estimate is this feature will be available quite soon in the next release candidate.

**Comment by nateshmbhat** [comment]
Any estimate on when the next release candidate can be released? The
library really wonderful. We want to use the above features as soon as
possible.

**Comment by avkonst** [comment]
implemented in Hookstate 4.0.0-rc7. Please try the functionality and let me know the feedback. Earlier implementation of this feature has been tested in production by a couple of companies. RC7 brings more support on top of it. Documentation updates are pending.

**Comment by avkonst** [comment]
This ticket is done. The documentation update will follow in https://github.com/avkonst/hookstate/issues/308
